src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs

```diff
@@ -1200,24 +1186,22 @@ public override int ExecuteNonQuery()
                     {
                         InternalExecuteNonQuery(completion: null, sendToPipe: false, timeout: CommandTimeout, out _);
                     }
+                    success = true;
                     return _rowsAffected;
                 }
                 catch (Exception ex)
                 {
-                    e = ex;
+                    diagnosticScope.SetException(ex);
+                    if (ex is SqlException sqlException)
+                    {
+                        sqlExceptionNumber = sqlException.Number;
+                    }
                     throw;
                 }
                 finally
                 {
                     SqlStatistics.StopTimer(statistics);
-                    if (e != null)
-                    {
-                        _diagnosticListener.WriteCommandError(operationId, this, _transaction, e);
-                    }
-                    else
-                    {
-                        _diagnosticListener.WriteCommandAfter(operationId, this, _transaction);
-                    }
+                    WriteEndExecuteEvent(success, sqlExceptionNumber, synchronous: true);
```

Comment:
The original code wasn't setting success or reporting the end execute event, so i added them for symmetry with ExecuteScalar and EcecuteReader.